### PR TITLE
SwiftGraphQL 3.0.0

### PR DESCRIPTION
**Pull request by maticzav**
This is a tracking PR for the next version of SwiftGraphQL. It focuses on the decoupling of query creation and query execution, extensible client, GraphQL WebSocket spec, SwiftUI bindings and caching mechanisms.

### CODE REVIEW

**Comment by paambaati** [comment]
@maticzav The minimum macOS version has been upgraded to 11.0 in https://github.com/maticzav/swift-graphql/commit/38925393134ad5af834d3557a2eb20c12d67deb6; is this a deliberate change? Can we keep supporting 10.15, as [Catalina is still considered a current release](https://en.wikipedia.org/wiki/MacOS_Catalina)?

File: Sources/GraphQLWebSocket/Client.swift [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+    
+    /// Sends a ping request and starts the response timeout.
+    private func ping() {
+        self.send(message: ClientMessage.ping())
+        self.emitter.send(Event.ping(received: false, payload: nil))
+        self.config.logger.debug("Emitted a PING message!")
+        
+        // We schedule a response timeout that has to be cleared
+        // in a timely manner by receiveing a new message from the server.
+        self.connectionDroppedTimer = Timer.scheduledTimer(
```
Comment (Amzd) [inline Sources/GraphQLWebSocket/Client.swift:375]: `ping()` is called every `config.keepAlive` seconds and each of these calls creates a new `connectionDroppedTimer` (destroying the current one) that waits for three times the keepAlive time. So only the last ping() call's `connectionDroppedTimer` can possibly ever return, is that intentional? 

Also I don't think it needs to repeat.

File: Sources/GraphQLWebSocket/Client.swift [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+    weak private var reconnectTimer: Timer?
+    
+    /// Timer that starts the disconnect procedure when there are no more listeners.
+    weak private var lazyCloseTimer: Timer?
+    
+    /// Timer that fires off when the server hasn't replyed with PONG for too long.
+    weak private var connectionDroppedTimer: Timer?
+    
+    /// Holds references to pipelines created by subscriptions. Each pipeline is identified by the
+    /// query id of the subscription that created a pipeline.
```
Comment (Amzd) [inline Sources/GraphQLWebSocket/Client.swift:131]:
```suggestion
    /// Holds references to pipelines created by subscriptions. Each pipeline is identified by the
    /// random id of the subscription that created a pipeline.
```
The id is a UUID() and not something that comes from the query

File: Sources/GraphQLWebSocket/Spec.swift [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    /// Returns a complete message for a connection with a given id.
+    ///
+    /// - parameter id: The ID of an established connection.
+    public static func complete(id: String) -> Self {
+        ClientMessage.complete(Complete(id: id))
+    }
+    
+    // MARK: - Interfaces
+    
+    public enum `Type`: String, CaseIterable, Codable {
```
[the remark this answers is not in the document]
Comment (maticzav) [reply Sources/GraphQLWebSocket/Spec.swift:84]: I like the suggestion, thanks!

File: Sources/GraphQLWebSocket/Client.swift [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+        self.health = .reconnecting(retry: retry + 1)
+        
+        self.config.logger.debug("Scheduling \(retry)-th connect retry!")
+        self.reconnectTimer = Timer.scheduledTimer(
+            withTimeInterval: TimeInterval(self.config.retryWait(retry)),
+            repeats: false,
+            block: { [weak self] _ in
+                // A reconnect should always happen if the client connects eagerly,
+                // even if we have no current listeners.
+                guard let self = self, !self.pipelines.isEmpty, self.config.behaviour != .eager else {
```
[the remark this answers is not in the document]
Comment (maticzav) [reply Sources/GraphQLWebSocket/Client.swift:321]: Good catch, thanks!

File: Sources/GraphQLWebSocket/Client.swift (hunk 2 above)
Comment (maticzav) [reply Sources/GraphQLWebSocket/Client.swift:131]: I see your point, but I think it's important to point out that this same `id` that's being used to identify the pipeline also appears in the `id` parameter of the server message.

File: Sources/GraphQLWebSocket/Client.swift (hunk 1 above)
Comment (maticzav) [reply Sources/GraphQLWebSocket/Client.swift:375]: `ping` is only called after the server and client haven't sent messages in a while (i.e. not every `config.keepAlive`). We want to fire (return) only when the timer times out and we haven't received any messages in the meantime.

Does that make sense?

Comment (Amzd) [reply Sources/GraphQLWebSocket/Client.swift:374]: I understand, I think I must have misunderstood when I reviewed.

**Pull request merged**